### ISSUE TITLE
feature request: standard way to communicate the local registry

### ISSUE THREAD

**Issue by nicks** [kubernetes/minikube#8101]
Many Kubernetes clusters support an insecure local registry, to make pushing images directly to the cluster fast (c.f., https://minikube.sigs.k8s.io/docs/handbook/registry/#enabling-insecure-registries). But every cluster documents it manually, and it's difficult for developer tooling to figure out that it exists.

I've been working with sig-cluster-lifecycle to define a standard for this.
KEP PR: kubernetes/enhancements#1757
KEP tracking issue: kubernetes/enhancements#1755

I'm posting this here so that Minikube contributors are aware of this proposal and help me understand if there are any Minikube-specific requirements. The proposal is really dependent on all the local cluster solutions agreeing to implement this. (I've filed similar issues against Kind, K3d, and Microk8s)

**Comment by medyagh** [comment]
thank you @nicks for bringing this to our attention, 
I wonder if you have already seen our docs on this: 

minikube provides many ways to push images, two of them are very fast, SSH and docker-env, u could see the the table comparison here https://minikube.sigs.k8s.io/docs/handbook/pushing/

**Comment by afbjorklund** [comment]
And then we don't even count our own (internal) way of fast provisioning, by replacing the entire storage part of the container runtime with a "preloaded" version so that we don't even have to load them...

But I don't think it is fair to say that minikube deploys a local registry (by default), what we do is to expose the container runtime on the node - or even allow logging in to the nodes directly with ssh...

---

It would be nice if there was a standard way to deploy a standard registry (the way it works in OpenShift), then minikube could just use this. I guess it also means provisioning storage for it ?

Then we could also avoid ugly hacks like tunneling port 5000 and such, like the current registry*. Maybe even avoid the "insecure" part (HTTP) and use standard HTTPS, but that's a stretch goal...

\* `gcr.io/google_containers/kube-registry-proxy:0.4`

Which has been **removed**: https://github.com/kubernetes/kubernetes/commit/d6918bbbc0402fc81a53479f4b61b836d7c33a29
> Remove unmaintained kube-registry-proxy support from gce kube-up.

https://github.com/kubernetes/kubernetes/tree/release-1.9/cluster/addons/registry

--> https://github.com/kubernetes-sigs/kubespray/tree/master/roles/kubernetes-apps/registry

----

I think we'll continue to use `docker-env`/`podman-env` (and `minikube cache`), with `ssh docker` and `ssh sudo podman` as fallbacks, and then go straight for a "real" k8s registry deployment ?

The others have more up-front issues, since they have decided to go with containerd and cri-o which means that they will **need** to have some kind of registry server running. Minikube does not.

**Comment by afbjorklund** [comment]
There is one other issue, and that is that even if you have a registry - you still can't _build_ images! The existing cloud provider container registries usually have some kind of "build service" attached...

**Comment by nicks** [comment]
Hi @medyagh and @afbjorklund ! Thanks for all that info. Ya, I've been consistently impressed by how many options minikube offers, and how well thought-out they are :smile: 

To set context on where I'm coming from: I work on Tilt. We spend a large amount of time educating our users on how to set up a cluster for local development. And there's a fair amount of code in Tilt that tries to do capability detection on clusters, so that Tilt can "auto-configure" if you already have a local registry set up. We don't want to put this configuration burden on users.

The point of the KEP is to settle on a standard protocol so that clusters can communicate that they support a local registry, and tools can consume this information and configure themselves.

The local clusters seem to be converging towards a local registry as a best practice, because the protocols are pretty standard by now and it has the best performance in the iterative development case (when you expect most of the base layers are already cached).

The questions I'd love your feedback on are:
1) If the KEP's proposal seems like a good way to communicate this info
2) If it's worth standardizing on how we communicate the other image-ingestion methods that minikube supports (either as part of this KEP or a follow-up KEP that builds on it)

Thanks!

**Comment by afbjorklund** [comment]
What would be really nice would be if someone would write a registry server, that would read and write images from the _same_ storage as the node uses for deploying new pods. So that one could push images to a server without having to pull (and duplicate) them before using, like it works today...

Before that is in place, and before multi-node is the default, it seems like overkill to deploy a registry server. Better to talk directly to the container runtime ? There is also some interesting work going on in k3s/k3c, on how to integrate BuildKit better with containerd. It's really a big hole in the CRI, this.

https://github.com/rancher/k3s
https://github.com/rancher/k3c

![k3s-k3c](https://user-images.githubusercontent.com/10364051/82371428-791d1e00-9a1a-11ea-9ae0-944cf5f830d3.png)

----

> The local clusters seem to be converging towards a local registry as a best practice, because the protocols are pretty standard by now and it has the best performance in the iterative development case (when you expect most of the base layers are already cached).

Running an insecure registry on port 5000 is really just half of a real registry, without benefits...
(tunneling localhost:5000 to get around the usual https requirement, to not have use a ssl cert)

But I still think it's a good idea to standardize how users can find their cluster-local registry server.
Just not the optimal solution for **local** development, since it takes more time and more effort ?


We are going to have to do _something_ once minikube goes multi-node, a registry might be it.

Another option is just looping over all the nodes, but that is less than ideal - for most scenarios.

**Comment by nicks** [comment]
In the KEP, I marked as out-of-scope speculation on what local image loaders might look like in the future. Mainly trying to describe what exists today, and provide some pattern for extending it with more options. 

Do you have a specific proposal for what additional image-loading options should be in the KEP?
https://github.com/kubernetes/enhancements/blob/c7119416a012ea51936a0727627b2c56b9973937/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry/README.md#the-local-image-loaders-configmap

I go back and forth on whether the combined image-builder/container-runtime architecture is going to end up being the dominant way of doing things. It has some significant upsides, but also significant downsides. That space seems frothy and unlikely to converge towards a single solution soon, so not sure how useful it is to wait on it.

**Comment by afbjorklund** [comment]
> Do you have a specific proposal for what additional image-loading options should be in the KEP?

Not really, what we discussed together with KIND was doing a more "portable" `docker load`...

In minikube we have this `cruntime` abstraction, and that allows us to load images.
Then it will call the appropriate command-line tool, for when `crictl` is **not** enough.

e.g. `docker load -i` or  `sudo podman load -i` or `sudo ctr -n=k8s.io images import`.
But it's not really available as a stand-alone tool, and thus there we only have "crictl pull".

So basically that was the capability, something to call on a tarball to load it into the runtime ?

----

The archive would have been previously created by `docker save` or similar library function...

I think the "kind load docker-image" is just a shortcut for doing both of those steps at once ?
i.e. `docker save foo | docker load` (but using two different daemons, I'm sure you understand)

We use it as a part of the `minikube cache`, to _avoid_ having to use a (local or remote) registry.

But we don't really have a one-off call, when you use minikube cache it will try to remember it.
That is it will be added to configuration and try to load it the next time as well, for that profile.

Also wonder if our current tool works for all drivers, it might be skipped for the "none" driver ?

**Comment by nicks** [comment]
ya, that makes me wonder if the real spec we need is something like

```
loadCommand: "docker load"
```

where loadCommand expects the image tarball on stdin, so that tools can figure out how to pipe it in from their own image store (whether it's docker or something else)

In any case, I'm doing a last call for any major objections or things I overlooked before we merge the initial spec for local registries:

https://github.com/kubernetes/enhancements/pull/1757
https://github.com/kubernetes/enhancements/blob/c5b6b632811c21ababa9e3565766b2d70614feec/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry/README.md#design-details

If you're OK with this proposal, a thumbs-up on the PR would be a help a lot! Thanks!

**Comment by nicks** [comment]
The KEP has been marked implementable! I suspect it will be straightforward to implement, but let me know if there's any way I can assist!

**Comment by afbjorklund** [comment]
(please ignore the link to 8494)
